### PR TITLE
Refactor query state to operate on a buffer

### PR DESCRIPTION
**Pull request by annevk**
- [x] At least two implementers are interested (and none opposed):
   * Firefox
   * Safari
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/pull/26158
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/master/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=1140965
   * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1671908
   * Safari: https://bugs.webkit.org/show_bug.cgi?id=218046

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

***
<a href="https://whatpr.org/url/558.html" title="Last updated on Nov 2, 2020, 9:43 AM UTC (38f6ae8)">Preview</a> | <a href="https://whatpr.org/url/558/a19495e...38f6ae8.html" title="Last updated on Nov 2, 2020, 9:43 AM UTC (38f6ae8)">Diff</a>

### CODE REVIEW

**Review by domenic — APPROVED** [review]
This seems to pass all the tests when implemented in jsdom/whatwg-url. It's more complicated, but I assume it's for a good reason related to #557...

**Comment by annevk** [comment]
Thanks, it's incomplete still as the percent-encoding algorithm will also need changes. I do think using a buffer coupled with https://github.com/whatwg/encoding/issues/235#issuecomment-712820914 gives us the "nicest" way of making this work, but happy to receive feedback otherwise. (The one weirdness is that we don't have a per-code-point percent-encode with arbitrary encoding anymore, but no caller really needs that. The UTF-8 caller can just pass its single code point as a string.)

**Comment by annevk** [comment]
The corresponding Encoding PR for the second commit in this PR hasn't landed yet and as such it'll fail Build, but it's still worth reviewing at this point I think, especially for non-editorial concerns.

**Comment by annevk** [comment]
I plan to rebase this Monday and make it pass Build now that https://github.com/whatwg/encoding/pull/238 has landed.

File: url.bs [hunk 1]
```diff
@@ -217,81 +217,70 @@ inclusive, and U+007E (~).
...
 
  <li><p>Return <var>output</var>.
 </ol>
 
 <p>To <dfn for="code point" id=utf-8-percent-encode>UTF-8 percent-encode</dfn> a
 <a for=/>code point</a> <var>codePoint</var> using a <var>percentEncodeSet</var>, return the result
-of running <a for="code point">percent-encode after encoding</a> with <a for=/>UTF-8</a>,
-<var>codePoint</var>, and <var>percentEncodeSet</var>.
+of running <a for=string>percent-encode after encoding</a> with <a for=/>UTF-8</a>,
+<var>codePoint</var> as <a for=/>string</a>, and <var>percentEncodeSet</var>.
```
Comment (domenic) [inline url.bs:278]:
```suggestion
<var>codePoint</var> as a <a for=/>string</a>, and <var>percentEncodeSet</var>.
```

File: url.bs [hunk 2]
```diff
@@ -217,81 +217,70 @@ inclusive, and U+007E (~).
...
 
 <p>To <dfn export for=string>UTF-8 percent-encode</dfn> a <a for=/>string</a> <var>input</var> using
 a <var>percentEncodeSet</var>, return the result of running
 <a for=string>percent-encode after encoding</a> with <a for=/>UTF-8</a>, <var>input</var>,
-<var>percentEncodeSet</var>, and false.
+<var>percentEncodeSet</var>.
```
Comment (domenic) [inline url.bs:283]:
```suggestion
and <var>percentEncodeSet</var>.
```

**Review by domenic — APPROVED** [review]
It looks like I can't really test this with jsdom/whatwg-url since we're still UTF-8 only. So my review of the second commit is just editorial.

Maybe I should sink some time into fixing that, but doing it properly would probably require a from-scratch implementation of the Encoding standard which exposes all of the spec-level primitives. Probably not something I can tackle any time soon...

**Comment by domenic** [comment]
Can we get web platform tests for this that are URL-specific? Probably not part of the JSON files though.

File: url.bs [hunk 3]
```diff
@@ -217,81 +217,70 @@ inclusive, and U+007E (~).
...
- <li><p>Let <var>bytes</var> be the result of <a lt=encode>encoding</a> <var>codePoint</var> using
- <var>encoding</var>.
+ <li><p>Let <var>encoder</var> be the result of <a>getting an encoder</a> from <var>encoding</var>.
 
- <li>
-  <p>If <var>bytes</var> starts with 0x26 (&amp;) 0x23 (#) and ends with 0x3B (;), then:
-
-  <ol>
-   <li><p>Let <var>output</var> be <var>bytes</var>, <a>isomorphic decoded</a>.
+ <li><p>Let <var>inputQueue</var> be <var>input</var> onverted to an <a for=/>I/O queue</a>.
```
Comment (andreubotella) [inline url.bs:227]:
```suggestion
 <li><p>Let <var>inputQueue</var> be <var>input</var> converted to an <a for=/>I/O queue</a>.
```

**Comment by annevk** [comment]
Thanks for the feedback! I put up some fairly basic tests over at https://github.com/web-platform-tests/wpt/pull/26317 but they should be easy to expand upon if we all think that format is workable.

**Pull request merged**